ace: d0ix implementation

tmleman opened the pull request:
D0i3 (dynamic power gating) implementation for meteorlake.

Under favorable conditions, the FW may try to enter D0i3 state. During that the memory will remain unchanged but core will be reset. That's why values of all necessary core registers have to be store in memory.

Magic value in lpsram will inform bootloader that core should go alternate boot vector. That will allow to restore previously stored core context.

tmleman committed:
intel: adsp: power down procedure rename

This patch is renaming primary core power down procedure by removing ACE
prefix. Prefix was added to avoid name conflicts while linking. The
problem is now resolved and name can be revert to the original version.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel: adsp: add restore vector after pg

This patch adds core restore vector. FW will restore core context from
memory and then return to the state it was in before power gaiting.

Adding struct of LPSRAM header. ROM during reboot check this struct in
serching for magic value and restore vector address.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel: adsp: enable d0i3 in device tree

This patch adds D0i3 definition to available core power states in device
tree.

Additionally, changing the name of the OFF state to better reflect what
is in the documentation.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,29 +26,194 @@
+	__asm__ volatile("rsr.vecbase %0"
+		: "=a"(vecbase));
+	return vecbase;
+}
+
+static ALWAYS_INLINE void write_vecbase(uint32_t vecbase)
+{
+	__asm__ volatile("wsr.vecbase %0"
+		: : "a"(vecbase) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave2(void)
+{
+	uint32_t excsave2;
+
+	__asm__ __volatile__("rsr.excsave2 %0"
+		: "=a"(excsave2));
+	return excsave2;
+}
+
+static ALWAYS_INLINE void write_excsave2(uint32_t excsave2)
+{
+	__asm__ __volatile__("wsr.excsave2 %0"
+		: : "a"(excsave2) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave3(void)
+{
+	uint32_t excsave3;
+
+	__asm__ __volatile__("rsr.excsave3 %0"
+		: "=a"(excsave3));
+	return excsave3;
+}
+
+static ALWAYS_INLINE void write_excsave3(uint32_t excsave3)
+{
+	__asm__ __volatile__("wsr.excsave3 %0"
+		: : "a"(excsave3) : "memory");
+}
```
@ceolin I have doubts about the location of this code, but I haven't found a better place. Any suggestions?

tmleman replied:
@ceolin alternatively, I can use macros **RSR** and **WSR** (from _arch/xtensa/arch_inlines.h_). I would only have to add support for user registers (_RUR.THREADPTR_).

ceolin replied:
The place is fine, these functions that don't look good. Can't you group them by functionality instead of per instruction ?
For example, you have `write_threadptr`, `write_vecbase`, `write_excsave2` and `write_excsave3` and they are used only once and together:
```
+		write_threadptr(core_desc[core_id].thread_ptr);
+		write_vecbase(core_desc[core_id].vecbase);
+		write_excsave2(core_desc[core_id].excsave2);
+		write_excsave3(core_desc[core_id].excsave3);
```

Couldn't they be just one function like `*save_context*()` ?

Same apply for other similar functions

tmleman replied:
I replaced this with usage of RSR/WSR macros. I think this looks better, less code added :)

tmleman commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -125,14 +223,24 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+
+		soc_cpus_active[cpu] = true;
+		z_xtensa_cache_flush_inv_all();
+		z_xt_ints_on(core_desc[cpu].intenable);
```
@ceolin this is big commit but I didn't find nice way to split it. Is it ok?

ceolin replied:
Yep, that is totally fine IMHO. You have split the changes into different commits.

tmleman committed:
arch: xtensa: add macros for user registers

This patch adds RUR and WUR macros to read and write user special
registers.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,29 +26,194 @@
+enum PowerFlow {
+	D0_TO_D0i3,
+	D0i3_TO_D0
+};
```
That looks really weird, can't you map the flow in relation to Zephyr pm states ? Or even better, just use `enum pm_state` ?

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,29 +26,194 @@
+
+void software_irq_handler(void *arg)
+{
+	uint32_t core_id = processor_id_reg();
```
ARG_UNUSED(arg);

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,29 +26,194 @@
+	z_xt_ints_off(0xffffffff);
+	IRQ_CONNECT(SOFTWARE_INTERRUPT_NUMBER, 0, software_irq_handler, 0, 0);
+	z_xtensa_irq_enable(SOFTWARE_INTERRUPT_NUMBER);
+	z_xt_set_intset(1<<SOFTWARE_INTERRUPT_NUMBER);
```
Why this is here ? That should be done once and not every time the system goes idle ...

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,29 +26,194 @@
+		lpsheader->lp_restore_vector = (uint32_t) &dsp_pg_restore;
+		core_desc[core_id].flow = D0i3_TO_D0;
+		z_xtensa_cache_flush_inv_all();
+		k_cpu_idle();
```
I am really confuse with this code, isn't this isr used to wake up a core ? Why this is being used when the core is going to be idle ? Shouldn't this be handled in the `power_gate()` function ?

How this is supposed to be used ? When this interruption is triggered ?

tmleman replied:
The idea behind isr was that we trigger software interrupt to force context switch and that way save core state into the memory. But for this moment I have gave it up. Maybe I'll come back to it later.

tmleman committed:
intel: adsp: d0i3 initial implementation

This patch adds initial implementation of dynamic power gating for ace
platforms. The code is limited to enabling PG in HW registers and
disabling all interrupts.

CPU context storing and restoring still needs to be implemented. PG at
this moment will cause FW to halt during reboot.

Patch contains small refractor. Value CORE_POWER_CHECK_DELAY has been
moved to common header so it can be reused in other places.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented:
@ceolin please revisit after update.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,17 +40,97 @@
+	"  wsr   a0, WINDOWBASE\n\t"
+	"  rsync\n\t");
+
+	{
```
Why this compound statement ?

tmleman replied:
The compiler sees _power_gate()_ as one, but it works like two separate functions. At this point we have never return from _k_cpu_idle()_. We get here from ROM because of this:

```c
lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
lpsheader->lp_restore_vector = (uint32_t) &dsp_pg_restore;
```

With enabled power gating, core during waiti goes down. Memory has not been changed but all processor registers have been reset. In this way, I force all variables used after the reset to have a smaller (local) scope.

ceolin replied:
Not following it, why do you need to force this variables to have a smaller scope ? The function returns at the end of this block, I think I am missing something.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,17 +40,97 @@
+		(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
+
+	xthal_window_spill();
+	core_desc[core_id].vecbase = RSR("VECBASE");
+	core_desc[core_id].excsave2 = RSR("EXCSAVE2");
+	core_desc[core_id].excsave3 = RSR("EXCSAVE3");
+	core_desc[core_id].thread_ptr = RUR("THREADPTR");
+	__asm__ volatile("mov %0, a0" : "=r"(core_desc[core_id].a0));
+	__asm__ volatile("mov %0, a1" : "=r"(core_desc[core_id].a1));
```
Why don't you put this in a function helper ? Something like:
`static void _save_core_context()` ? It makes this function more readable, since one can follow the logic and not all these assembly lines.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,17 +40,97 @@
+	"  movi  a1, 1\n\t"
+	"  wsr   a1, WINDOWSTART\n\t"
+	"  wsr   a0, WINDOWBASE\n\t"
+	"  rsync\n\t");
```
I don't think that is a good place to define it. Maybe put it in `power_down.S` ?

tmleman replied:
I would have to write everything in assembly.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,17 +40,97 @@
+};
+
+static struct core_state core_desc[CONFIG_MP_NUM_CPUS] = { 0 };
+static struct core_state *desc_alias = &core_desc;
```
What is the purpose of this alias ?

tmleman replied:
I have not found a better solution. I'm using it to get struct address after core reset.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -5,16 +5,30 @@
 
 #include <ace_v1x-regs.h>
 
+#define CHECK_DELAY             256
```
Is this a random or documented value ? It is worth some comment.

ceolin replied:
ok, so doesn't make sense to put in a header instead of having it duplicated ? It seems a quite random value, so at least, lets keep it in on single place.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -5,16 +5,30 @@
+#define LPSCTL_BATTR_MASK       BIT_MASK_FROM_RANGE(12, 16)
 #define SRAM_ALIAS_BASE         0xA0000000
 #define SRAM_ALIAS_MASK         0xF0000000
+#define MEMCTL_INIT_BIT         23
```
Why not `#define MEMCTL_INIT_BIT        BIT(23)` ?
And then in the next line you just use the macro ?

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -5,16 +5,30 @@
+#define L2_INTERRUPT_MASK       (1<<L2_INTERRUPT_NUMBER)
+
+#define L3_INTERRUPT_NUMBER     6
+#define L3_INTERRUPT_MASK       (1<<L3_INTERRUPT_NUMBER)
```
Can't these interruptions come from dt ?

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,17 +40,97 @@
+	"  rsync\n\t");
+
+	{
+		core_id = arch_proc_id();
```
Why overriding this parameter ?

tmleman replied:
I'm not overriding it, it's more like reinitialization. Reason for that has been given in one of the earlier comments.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -26,17 +40,97 @@
+
+		__asm__ volatile("l32i a14, a13, 0x14\n\t");
+		WUR("THREADPTR", value);
+	}
```
same thing about having it in a separated function, `_restore_core_context()`

mwasko commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -16,25 +16,31 @@
 	power-states {
+		suspend: suspend {
+			compatible = "zephyr,power-state";
+			power-state-name = "suspend-to-idle";
```
ACE D0i3 power state does not match suspend-to-idle and it can be confusing. The recommendation is to use substate_id set to ACE defined D0i3 value.

tmleman replied:
Introducing substates currently is not necessary. I have change states name to better reflect those defined in documentation.

ceolin commented on include/zephyr/sys/util_macro.h:
```diff
@@ -73,6 +73,15 @@ extern "C" {
+ * @param to  The most significant bit
+ */
+#define BIT_MASK_FROM_RANGE(from, to) ((0xFFFFFFFF << (from)) & \
+	(0xFFFFFFFF >> (sizeof(uint32_t) * 8 - 1 - (to))))
```
You don't need to define it. You can use `GENMASK` can't you ?

tmleman commented:
fyi @mmaka1
